zpeclient.go

```
@@ -0,0 +1,227 @@
+)
+
+// roleMappings - cr cache
+type roleMappings struct {
+	roleToPrincipals map[string][]*simplePrincipal
+	roleToAssertion  map[string][]*simpleAssertion
+}
+
+// simplePrincipal - principal data
+type simplePrincipal struct {
+	memberRegex *regexp.Regexp
+	expiration  time.Time
+}
+
+// simpleAssertion - processed policy
+type simpleAssertion struct {
+	resource *regexp.Regexp
+	action   *regexp.Regexp
+	effect   string
+}
+
+// Cache - cache for athenzdomains CR
+type Cache struct {
+	CrIndexInformer cache.SharedIndexInformer
+	DomainMap       map[string]roleMappings
+	lock            sync.RWMutex
+}
+
+// BuildCache - generate new athenzdomains cr cache
+func BuildCache(crIndexInformer cache.SharedIndexInformer) {
+	domainMap := make(map[string]roleMappings)
+	privateCache = &Cache{
+		CrIndexInformer: crIndexInformer,
+		DomainMap:       domainMap,
+	}
+	crIndexInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			item, ok := obj.(*v1.AthenzDomain)
+			if !ok {
+				fmt.Println("Unable to convert informer store item into AthenzDomain type.")
+				return
+			}
+			privateCache.addObj(item)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			newItem, ok := newObj.(*v1.AthenzDomain)
+			if !ok {
+				fmt.Println("Unable to convert informer store item into AthenzDomain type.")
+				return
+			}
+			privateCache.updateObj(newItem)
+		},
+		DeleteFunc: func(obj interface{}) {
+			item, ok := obj.(*v1.AthenzDomain)
+			if !ok {
+				fmt.Println("Unable to convert informer store item into AthenzDomain type.")
+				return
+			}
+			privateCache.deleteObj(item)
+		},
+	})
+}
+
+// parseData - helper function to parse AthenzDomain data and store it in domainMap
+func parseData(domainMap map[string]roleMappings, domainName string, item *v1.AthenzDomain) error {
+	crMap := domainMap[domainName]
+	if item == nil || item.Spec.SignedDomain.Domain == nil {
+		return errors.New("Some required fields are nil")
```

Can we return an error saying which required fields are nil?